### PR TITLE
Allow `csv` module to support both `str` and `unicode` types in Python 2

### PR DESCRIPTION
**Pull request by rowillia**
Upgrading typeshed broke some of our code that was using `DictReader` and `unicode_literals`

### CODE REVIEW

File: stdlib/2and3/csv.pyi [hunk 1]
```diff
@@ -22,25 +22,25 @@ _Dialect = Union[str, Dialect]
 _DictRow = Dict[str, Any]
 
 class Dialect(object):
-    delimiter = ...  # type: str
-    quotechar = ...  # type: Optional[str]
-    escapechar = ...  # type: Optional[str]
+    delimiter = ...  # type: Text
```
Comment (JelleZijlstra) [inline stdlib/2and3/csv.pyi]: This kind of thing can also break working code if people rely on `Dialect.delimiter` actually being a `str` in Python 2. Maybe these attributes need to be Any or str, unfortunately.

Does your code work if you use `Text` just in argument types?

File: stdlib/2and3/csv.pyi [hunk 2]
```diff
@@ -65,33 +65,33 @@ if sys.version_info >= (3, 6):
...
-        def __init__(self, f: Iterator[str], fieldnames: Sequence[str] = ...,
-                     restkey: Optional[str] = ..., restval: Optional[str] = ..., dialect: _Dialect = ...,
+        fieldnames = ...  # type: Sequence[Text]
+        def __init__(self, f: Iterator[Text], fieldnames: Sequence[Text] = ...,
+                     restkey: Optional[Text] = ..., restval: Optional[Text] = ..., dialect: _Dialect = ...,
                      *args: Any, **kwds: Any) -> None: ...
-        def __iter__(self) -> Iterator[OrderedDict[str, str]]: ...
-        def next(self) -> OrderedDict[str, str]: ...
+        def __iter__(self) -> Iterator[OrderedDict[Text, Text]]: ...
+        def next(self) -> OrderedDict[Text, Text]: ...
```
Comment (JelleZijlstra) [inline stdlib/2and3/csv.pyi]: Can this really return unicode in Python 2? I haven't tested it.

File: stdlib/2and3/csv.pyi (hunk 1 above)
Comment (rowillia) [inline stdlib/2and3/csv.pyi]: This is a case where this breaks with `unicode_literals`.  I know `unicode_literals` has effectively been deprecated, but we still have a ton of code that makes use of it and we can't rip it out before upgrading to Python 3 without risking further breakage.

I suppose all of these fields could be modeled with as a `@property` who accepts a `Union[str, Text]` for the setter and return a `str`

**Comment by gvanrossum** [comment]
(I don't know that it's officially been deprecated -- it's just that I
personally dislike it, at least in part because it wreaks havoc with
mypy...)

**Comment by JelleZijlstra** [comment]
@gvanrossum would you mind checking if this PR introduces any problems in your codebase? This PR introduces a little white lie, but hopefully it won't cause problems in real code.

**Comment by gvanrossum** [comment]
Well, I found one:
```
Argument 1 to "writerow" of "DictWriter" has incompatible type OrderedDict[unicode, str]; expected Dict[str, Any]
```
Here the argument was read from a `DictReader`, but I expect that there's a general issue now due to variance, so you can't just change the signature of `writerow()` I suspect.

**Comment by JelleZijlstra** [comment]
Thanks!

I'm afraid we may have to fall back to making the key type `Any` then.

**Comment by rowillia** [comment]
Thanks for taking a look @gvanrossum !  At least we'd still have `str` types in Py3

**Comment by rowillia** [comment]
@JelleZijlstra @gvanrossum changed the key type of `DictReader` to `Any` in Python 2

**Pull request merged**